plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -221,16 +219,49 @@ func findNodesThatFit(
 }
 
 // Checks whether node with a given name and NodeInfo satisfies all predicateFuncs.
-func podFitsOnNode(pod *v1.Pod, meta interface{}, info *schedulercache.NodeInfo, predicateFuncs map[string]algorithm.FitPredicate) (bool, []algorithm.PredicateFailureReason, error) {
-	var failedPredicates []algorithm.PredicateFailureReason
-	for _, predicate := range predicateFuncs {
-		fit, reasons, err := predicate(pod, meta, info)
-		if err != nil {
-			err := fmt.Errorf("SchedulerPredicates failed due to %v, which is unexpected.", err)
-			return false, []algorithm.PredicateFailureReason{}, err
+func podFitsOnNode(pod *v1.Pod, meta interface{}, info *schedulercache.NodeInfo, predicateFuncs map[string]algorithm.FitPredicate,
+	ecache *EquivalenceCache) (bool, []algorithm.PredicateFailureReason, error) {
+	var (
+		equivalenceHash  uint64
+		failedPredicates []algorithm.PredicateFailureReason
+		needDoPredicate  bool
+		eCacheAvailable  bool
+	)
+	if ecache != nil {
+		// getHashEquivalencePod will return immediately if no equivalence pod found
+		equivalenceHash = ecache.getHashEquivalencePod(pod)
+	}
+	for predicateKey, predicate := range predicateFuncs {
+		// If equivalenceCache is available
+		if eCacheAvailable = (ecache != nil && equivalenceHash != 0); eCacheAvailable {
+			// PredicateWithECache will returns it's cached predicate results
+			fit, reasons, invalid := ecache.PredicateWithECache(pod, info.Node().GetName(), predicateKey, equivalenceHash)
+			if !invalid {
+				if !fit {
+					failedPredicates = append(failedPredicates, reasons...)
+				}
+			} else {
+				needDoPredicate = true
+			}
+		} else {
+			needDoPredicate = true
 		}
-		if !fit {
-			failedPredicates = append(failedPredicates, reasons...)
+
+		// we need to execute predicate functions as normal since equivalence cache does not work
+		if needDoPredicate {
```

Comment:
(optional, but nice to have)i like the comment, but better yet, sugggest separate function here `processPredicateFunctions()` ?  Might be more readable in the future, and help highlight that the equivalence cache is a one-off optimization that runs before the normal behaviour.